proposals/dark_mode_extension_icons.md

```diff
@@ -0,0 +1,200 @@
+Firefox can specify icons).
+
+### Known Consumers
+
+The Chromium bug has a significant amount of developer interest.
+
+## Specification
+
+### Schema
+
+manifest.json
+```
+"icon_variants": [
+  {
+    "any": "any.svg",
+  },
+  {
+    "16": "16.png",
+    "32": "32.png",
+  },
+  {
+    "16": "dark16.png",
+    "32": "dark32.png",
+    "color_scheme": "dark",
+  },
+  {
+    "16": "light16.png",
+    "32": "light32.png",
+    "color_scheme": ["dark", "light"]
+  }
+]
+```
+
+Set icon_variants dynamically.
+```
+const exampleProperties: {string: string | ImageData}[] = [
+  {
+    "any": "any.svg",
+  },
+  {
+    "16": "16.png",
+    "32": "32.png"
+  },
+  {
+    "16": darkImageData16,
+    "32": darkImageData32,
+    "color_scheme": "dark"
+  },
+  {
+    "16": lightImageData16,
+    "32": lightImageData32,
+    "color_scheme": "light"
+  }
+];
+
+// action.setIcon().
+const createProperties = {variants: exampleProperties};
+action.setIcon(createProperties);
+
+// menus.*(), for supporting browsers.
+const menusProperties = {icon_variants: exampleProperties};
+menus.create(menusProperties);
+menus.update(menusProperties);
+```
+
+A benefit of this new structure is that it's more resiliant to future changes, thus allowing for more keys such as density (e.g. 2dppx), purpose (e.g.
+monochrome), and etc.
+
+## setIcon()
+Incumbent action.setIcon(), for reference.
+```
+action.setIcon({
+  path,
+  tabId
+  imageData,
+});
+```
+
+### Behavior
+
+Existing manifest key. The behavior of the icons manifest key will remain unchanged.
+```
+{
+  "icons": {
+    "64": "icon_64.png"
+  },
+  "action": {
+    "default_icon": {
+      "64": "action_64.png"
+    }
+  }
+}
+```
+
+Order of precedence. The new `icon_variants` keys and subkeys will take
+precedent, followed by the incumbent `icons` key.
+
+The `dark` icon variant is used when a dark color scheme is used for the
+surrounding UI, often referred to as "dark mode".
+The `light` icon variant is used otherwise.
+
+### New Permissions
+
+N/A.
+
+### Manifest File Changes
+
+1. If icon_variants are supplied, icons can be ignored.
+1. If icon_variants contain an icon object with duplicate conditions, the first match will be
+used.
+1. Any icon object that does not contain a "color_scheme" key will apply to both
+light and dark.
+1. If only one icon object is supplied, it will be used for both light and dark.
+1. icon_variants will not cause an error in the event that it is invalid
+Instead, it can be ignored altogether or just emit a warning. Warnings are
+preferred over errors because they're more adaptable to changes in the future.
+1. Neither "dark" nor "light" color_scheme's are required.
+1. Icon objects missing the color_scheme property will apply to light and dark
+mode. It could also be explicity set using {"color_scheme": ["dark", "light"]}.
+1. If icon_variants are provided, the top level icons key and
+action.default_icon will be ignored. There is no fallback.
```

Comment:
Considering the `icon_variants` being present. `action.default_icon` being set and `action.icon_variants` missing.  As follows:
```json
{
  "icon_variants": [...],
  "action": {
    "default_icon": "...path"
  }
}
```

Intuitively it feels `action.default_icon` should override whatever is set in `icon_variants`. Same for setting [`action.theme_icons` ](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/action) would override what is set in the main `icon_variants`.

My proposal would be:
```suggestion
If the top level icon_variants key is provided, the top level icons key will be ignored.
For action icons follows a strict order of precedence:
action.icon_variants > action.theme_icons > action.default_icon > icon_variants > icons
Any programmatically set icons using `action.setIcon` would override the declaratively defined icons mentioned above.
```